fuse_constraints/include/fuse_constraints/fixed_3d_landmark_cost_functor.h

```diff
@@ -0,0 +1,206 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Author: Oscar Mendez
+ *  Created on Mon Nov 12 2023
+ *
+ *  Copyright (c) 2023, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_CONSTRAINTS_FIXED_3D_LANDMARK_COST_FUNCTOR_H
+#define FUSE_CONSTRAINTS_FIXED_3D_LANDMARK_COST_FUNCTOR_H
+
+#include <fuse_constraints/normal_prior_orientation_3d_cost_functor.h>
+#include <fuse_core/eigen.h>
+#include <fuse_core/fuse_macros.h>
+#include <fuse_core/util.h>
+
+#include <Eigen/Core>
+
+namespace fuse_constraints
+{
+
+/**
+ * @brief Create a prior cost function on the marker position, minimising reprojection error.
+ *
+ * The Ceres::NormalPrior cost function only supports a single variable. This is a convenience cost function that
+ * applies a prior constraint on the 3D position, orientation and calibration variables at once.
+ *
+ * The cost function is of the form:
+ *
+ *   cost(x) = || A * (K * [R_q | p] * [R_{b(3:6)} | b(0:2))] * X) - x  ||
+ *
+ * where, the matrix A and the vector b are fixed, p is the camera position variable, and q is the camera orientation
+ * variable, K is the calibration matrix created from the calibration variable, X is the set of marker 3D points,
+ * R_b(0:3) is the Rotation matrix from the fixed landmark orentation (b(3:6)), b(0:2) is the fixed landmark position
+ * and x is the 2D ovservations.
```

Comment:
Ignore the squaring comment, apologies.

Reply:
I think you are right, I'll change it, but I need to work out how this will work. 

The weighting matrix should apply to all the residuals, and would weight each reprojection error individually, but I suspect we won't get something like that directly at runtime. Instead, I assume our weights will come from the landmark pose covariance, rather than the observations?

So basically, i agree, but I need to work out how to calculate A from the covariance matrix. I've been looking at [this document](https://arxiv.org/pdf/2103.15980.pdf) for the 3D constraints, and on p.59 there are some starting points for the jacobians, but I need to sit down and think it through properly.